localstack-core/localstack/runtime/runtime.py

```diff
@@ -0,0 +1,213 @@
+        LOG.debug("[shutdown] Completed, bye!")
+
+    def _run_ready_monitor(self):
+        self._wait_for_gateway()
+        self._on_ready()
+
+    def _wait_for_gateway(self):
+        host_and_port = self.config.GATEWAY_LISTEN[0]
+
+        if not sync.poll_condition(
+            lambda: net.is_port_open(host_and_port.port), timeout=15, interval=0.3
+        ):
+            if LOG.isEnabledFor(logging.DEBUG):
+                # make another call with quiet=False to print detailed error logs
+                net.is_port_open(host_and_port.port, quiet=False)
+            raise TimeoutError(f"gave up waiting for gateway server to start on {host_and_port}")
+
+    def _clear_tmp_directory(self):
+        if self.config.CLEAR_TMP_FOLDER:
+            # try to clear temp dir on startup
+            try:
+                files.rm_rf(self.config.dirs.tmp)
+            except PermissionError as e:
+                LOG.error(
+                    "unable to delete temp folder %s: %s, please delete manually or you will "
+                    "keep seeing these errors.",
+                    self.config.dirs.tmp,
+                    e,
+                )
+
+    def _cleanup_resources(self):
+        threads.cleanup_threads_and_processes()
+        self._clear_tmp_directory()
+
+    # more legacy compatibility code
+    @property
+    def exit_code(self):
+        # FIXME: legacy compatibility code
+        from localstack.runtime import legacy
+
+        return legacy.EXIT_CODE.get()
+
+    @exit_code.setter
+    def exit_code(self, value):
+        # FIXME: legacy compatibility code
+        from localstack.runtime import legacy
+
+        legacy.EXIT_CODE.set(value)
+
+    def _run_shutdown_monitor(self):
+        # FIXME: legacy compatibility code. this can be removed once we replace access to the
+        #  ``SHUTDOWN_INFRA`` event with ``get_current_runtime().shutdown()``.
+        from localstack.runtime import legacy
+
+        legacy.SHUTDOWN_INFRA.wait()
+        self.shutdown()
+
+
+def create_from_environment() -> LocalstackRuntime:
+    """
+    Creates a new runtime instance from the current environment. It uses a plugin manager to resolve the
+    necessary components from the ``localstack.runtime.components`` plugin namespace to start the runtime.
+
+    TODO: perhaps we could control which components should be instantiated with a config variable/constant
+
+    :return: a new LocalstackRuntime instance
+    """
+    hooks.on_runtime_create.run()
+
+    plugin_manager = PluginManager(Components.namespace)
+    components = plugin_manager.load_all()
+
+    if not components:
+        raise ValueError(
+            f"No component plugins found in namespace {Components.namespace}. Are entry points created "
+            f"correctly?"
+        )
+
+    if len(components) > 1:
+        LOG.warning(
+            "There are more than one component plugins, using the first one which is %s",
+            components[0].name,
+        )
```

Comment:
nit: Feels a bit too lenient for me. I'd personally prefer it to raise in that case. :man_shrugging: